src/libraries/Microsoft.Extensions.Configuration/src/InternalConfigurationRootExtensions.cs

```diff
@@ -20,11 +20,16 @@ internal static class InternalConfigurationRootExtensions
         /// <returns>Immediate children sub-sections of section specified by key.</returns>
         internal static IEnumerable<IConfigurationSection> GetChildrenImplementation(this IConfigurationRoot root, string? path)
         {
-            return root.Providers
+            using ReferenceCountedProviders? reference = (root as ConfigurationManager)?.GetProvidersReference();
```

Comment:
I wonder if this code makes sense here. What if the "special" logic here was moved into `ConfigurationManager` - calling `GetProvidersReference()` and `ToList()`? We could still share the inner logic here, we just wouldn't be paying the cost nor complexity in code paths outside of ConfigurationManager.

This would also remove the need for the `internal GetProvidersReference()` method on ConfigurationManager.

Reply:
I initially went that route, but the problem is that `ConfigurationSection` also calls into this internal method and should use `ReferenceCountedProviders` when paired with `ConfigurationManager`.

I considered making a custom `IConfigurationSection` implementation for `ConfigurationManager` but was concerned that changing the runtime type from one public type to something else could be breaking to applications that might cast the `IConfigurationSection` to `ConfigurationSection`. Maybe that's too paranoid, but I would like to get this serviced.

I also considered doing this logic just in `ConfigurationManager.GetChildren()` and `ConfigurationSection.GetChildren()`, but at that point `ConfigurationRoot.GetChildren()` would be the only thing without this logic.

I could skip the `.ToList()` when this isn't being used with a `ConfigurationManager` though.

Reply:
> that ConfigurationSection also calls into this internal method and should use ReferenceCountedProviders when paired with ConfigurationManager.

Ah, I missed this. Good point. Thanks for clarifying.